transmogrifier/models.py

```diff
@@ -43,11 +49,26 @@ def not_empty(
         raise ValueError(message)
 
 
+@define
+class ListField:
+    def __hash__(self) -> int:
+        """Hash method to create unique identifier for Location objects."""
+        values = tuple(
+            [
+                tuple(attrib) if isinstance(attrib, list) else attrib
+                for attrib in attrs.astuple(self)
+            ]
+        )
+        return hash(values)
+
+
 @define
 class AlternateTitle:
     value: str = field(validator=instance_of(str))  # Required subfield
     kind: str | None = field(default=None, validator=optional(instance_of(str)))
 
+    __hash__ = ListField.__hash__
```

Comment:
It does feel off to define a class and not use it for inheritance, I think it could be worth a huddle to discuss this further

Reply:
And if the class isn't used for inheritance, is it better to just defining a `timdex_object_hash()` function and setting `__hash__ = timdex_object_hash()`? Not fully sure of the implications of that though

Reply:
Per [this comment](https://github.com/MITLibraries/transmogrifier/pull/206#discussion_r1714182188) above, I wonder if it's passing because the hash created is _always_ unique, so while no runtime exceptions, it's not actually deduping anything?

Reply:
Circled back to this after doing some more reading (this [thread](https://stackoverflow.com/questions/52390576/how-can-i-make-a-python-dataclass-hashable) was particularly helpful), I think either `unsafe_hash=True` or `timdex_object_hash()` works here, the `ListField` class feels like too much overhead if it can't properly be an inherited class due to the hash issues

Reply:
Thank you both for looking into this more closely! I agree about the `ListField` class feeling redundant as it's not accomplishing inheritance in the way we expected it to. 🤔 

Just wanted to note a couple items:

Regarding @ghukill 's initial comment: 
* The child classes do not inherit the `__hash__` method from the parent class `ListField` because `attrs` default behavior will write an `__eq__` method for the class. the Python 3 docs, it reads:
   > A class that overrides [__eq__()](https://docs.python.org/3/reference/datamodel.html#object.__eq__) and does not define [__hash__()](https://docs.python.org/3/reference/datamodel.html#object.__hash__) will have its [__hash__()](https://docs.python.org/3/reference/datamodel.html#object.__hash__) implicitly set to None. When the [__hash__()](https://docs.python.org/3/reference/datamodel.html#object.__hash__) method of a class is None, instances of the class will raise an appropriate [TypeError](https://docs.python.org/3/library/exceptions.html#TypeError) when a program attempts to retrieve their hash value. . .
   I believe this is why we can't simply inherit the `ListField` class. 🤔 
* `@define(frozen=True)` will make the list field (it decorates) **immutable**. The cause of all those errors cropping up when this setting is configured are field methods that involve creating the instance of the object and then proceeding to update attributes (e.g. [dates field method in Ead transform](https://github.com/MITLibraries/transmogrifier/blob/main/transmogrifier/sources/xml/ead.py#L257-L263)). Given the current state of our transforms, I agree that this approach is not feasible.

To be continued ...

Reply:
I'd like to clarify my comment a bit here after a discussion yesterday with @jonavellecuerdo.

I had not considered in this toy example that different classes (e.g. `DogC` vs `DogD`) _should_ likely not be equal, given they are different classes.  

While I was focused on instance equality vs hash method equality... @jonavellecuerdo observed that the class type should be a consideration.

Example, the following should _not_ be equal:
```python
AlternateTitle(value="x", kind="y") == Location(value="x", kind="y")
```

If we are thinking only of field names + values, then they _would_ be considered equal.  But it's clear that to dedupe, the class type is an important variable.  I got the impression that @jonavellecuerdo was exploring changes to reflect this, so won't dig in more here, but did want to include these noets that stemmed from that conversation yesterday.

Reply:
@ghukill Thank you for providing this summary! TLDR: From https://github.com/MITLibraries/transmogrifier/pull/206#discussion_r1716965905, I noticed that what `unsafe_hash=True` did _differently_ (at least prior to the updates introduced https://github.com/MITLibraries/transmogrifier/pull/206/commits/f81475b90ba01921858358d1f3f58a8f556110ef) is that it took _class types_ into account when generating the hash method. 

Prior to the updates in https://github.com/MITLibraries/transmogrifier/pull/206/commits/f81475b90ba01921858358d1f3f58a8f556110ef, this would've occurred:
```python
AlternateTitle(value="x", kind="y") == Location(value="x", kind="y") # returns True!
```
To fix this, the [class type is now taken into account my `timdex_object_hash`](https://github.com/MITLibraries/transmogrifier/pull/206/commits/f81475b90ba01921858358d1f3f58a8f556110ef#diff-62380015f300f79c3d62063c1dffd2c085e0ea5a1cfbf24946e5bb31c36d587c). [Unit tests](https://github.com/MITLibraries/transmogrifier/pull/206/commits/f81475b90ba01921858358d1f3f58a8f556110ef#diff-0adb667cd397bd56edce10eec11e1b10821740a10d72bd148b09645fc9108968R376-R397) to assert that: [if `x == y`, it must follow that `hash(x) == y` (and vice versa)](https://www.attrs.org/en/stable/hashing.html) have also been added.